pages/instructorAssessmentGroups/instructorAssessmentGroups.js

```diff
@@ -146,67 +158,112 @@ router.post('/', function(req, res, next) {
             if (ERR(err, next)) return;
             res.redirect(req.originalUrl);
         });
-    } else if (req.body.__action == 'addmember') {
+    } else if (req.body.__action == 'add_member') {
         const assessment_id = res.locals.assessment.id;
         const gid = req.body.gid;
-        const uids = req.body.addmemberuids;
+        const uids = req.body.add_member_uids;
         const uidlist = uids.split(/[ ,]+/);
         let failedUids = '';
         res.locals.errormsg = '';
-        (async () => {
-            for (const uid of uidlist) {
-                const params = [
-                    assessment_id,
-                    gid,
-                    uid,
-                    res.locals.authn_user.user_id,
-                ];
-                try {
-                    await sqldb.callAsync('assessment_groups_add_member', params);
-                } catch (err) {
-                    failedUids += '[' + uid + '] ';
-                }
-            }
-            if (failedUids.length > 0) {
-                res.locals.errormsg += 'Failed to add ' + failedUids + 'to Group No.' + gid + '. Please check if the uid exist.\n';
+
+        //verify the group belong to the assessment first
+        let params = {
+            assessment_id,
+            gid,
+        };
+        sqldb.query(sql.verify_group, params, function(err, result) {
```

Comment:
While it might be ok here, in general I think this pattern of "verify and then change" is risky because it tends to introduce [TOCTAU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) race conditions.

Instead, here I think it would be better to send the `assessment_id` to `assessment_groups_add_member()` so that we can check at the same time as we make the change. The checking should happen either as we lock the `groups` row, or _after_ we've locked the `groups` row.

(and similarly for the other functions like `_delete_member` and `_delete_group`)